frame/bags-list/src/list/mod.rs

```diff
@@ -0,0 +1,670 @@
+/// function behaves as if it does.
+pub(crate) fn notional_bag_for<T: Config>(weight: VoteWeight) -> VoteWeight {
+	let thresholds = T::BagThresholds::get();
+	let idx = thresholds.partition_point(|&threshold| weight > threshold);
+	thresholds.get(idx).copied().unwrap_or(VoteWeight::MAX)
+}
+
+/// Data structure providing efficient mostly-accurate selection of the top N id by `VoteWeight`.
+///
+/// It's implemented as a set of linked lists. Each linked list comprises a bag of ids of
+/// arbitrary and unbounded length, all having a vote weight within a particular constant range.
+/// This structure means that ids can be added and removed in `O(1)` time.
+///
+/// Iteration is accomplished by chaining the iteration of each bag, from greatest to least. While
+/// the users within any particular bag are sorted in an entirely arbitrary order, the overall vote
+/// weight decreases as successive bags are reached. This means that it is valid to truncate
+/// iteration at any desired point; only those ids in the lowest bag can be excluded. This
+/// satisfies both the desire for fairness and the requirement for efficiency.
+pub struct List<T: Config>(PhantomData<T>);
+
+impl<T: Config> List<T> {
+	/// Remove all data associated with the list from storage.
+	pub(crate) fn clear() {
+		crate::CounterForListNodes::<T>::kill();
+		crate::ListBags::<T>::remove_all(None);
+		crate::ListNodes::<T>::remove_all(None);
+	}
+
+	/// Regenerate all of the data from the given ids.
+	///
+	/// This is expensive and should only ever be performed during a migration, or when
+	/// the data needs to be generated from scratch again.
+	///
+	/// This may or may not need to be called at genesis as well, based on the configuration of the
+	/// pallet using this `List`.
+	///
+	/// Returns the number of ids migrated.
+	pub fn regenerate(
+		all: impl IntoIterator<Item = T::AccountId>,
+		weight_of: Box<dyn Fn(&T::AccountId) -> VoteWeight>,
+	) -> u32 {
+		Self::clear();
+		Self::insert_many(all, weight_of)
+	}
+
+	/// Migrate the list from one set of thresholds to another.
+	///
+	/// This should only be called as part of an intentional migration; it's fairly expensive.
+	///
+	/// Returns the number of accounts affected.
+	///
+	/// Preconditions:
+	///
+	/// - `old_thresholds` is the previous list of thresholds.
+	/// - All `bag_upper` currently in storage are members of `old_thresholds`.
+	/// - `T::BagThresholds` has already been updated and is the new set of thresholds.
+	///
+	/// Postconditions:
+	///
+	/// - All `bag_upper` currently in storage are members of `T::BagThresholds`.
+	/// - No id is changed unless required to by the difference between the old threshold list and
+	///   the new.
+	/// - ids whose bags change at all are implicitly rebagged into the appropriate bag in the new
+	///   threshold set.
+	#[allow(dead_code)]
+	pub fn migrate(old_thresholds: &[VoteWeight]) -> u32 {
+		// we can't check all preconditions, but we can check one
+		debug_assert!(
```

Comment:
Should also check that all `ListNodes::<T>::iter().map(|n| n.bag_upper)` are within `old_thresholds`.

Reply:
597e1865a7f8c509b56977428a60be8a9b2d4df1